### DISCUSSION TITLE
Looking at context names and chat IDs in more detail

### DISCUSSION THREAD

**Discussion by fjuma** [wise-agents/wise-agents#325] [General]
I looked into our usage of context name and chat IDs.

Currently, chat IDs are generated and used for the following cases:

1) Sequential coordination (an ID is generated for each request the coordinator receives and is used to keep track of the sequence of agents for the request and where to route the final response to)

2) Phased coordination (an ID is generated for each request the coordinator receives and is used to keep track of the agents required for each phase, where to route the final response to, and to keep track of the conversation history for the request)

3) Assistant agent (a single ID is generated and used for all requests to keep track of the conversation history)

4) `LLMWiseAgentWithTools` (an ID is generated for each request the agent receives and is used to keep track of the tool calls needed for the request and the output of the tools).

Currently, a `WiseAgentContext` has a name and can be associated with multiple requests so the context contains dictionaries that map chat IDs to various values.

For example, if we have an `AssistantAgent` that can send a message to a `SequentialCoordinator` that invokes agents `A` and `B`, we'd have:

* a context name
* a single chat ID generated by the `AssistantAgent`
* a chat ID generated by the `SequentialCoordinator` to process each request it receives 

If we have a `PassThroughClientAgent` that sends a message to a `SequentialCoordinator`, we'd have:

* a context name
* a chat ID generated by the `SequentialCoordinator` to process each request it receives

Basically, we can see that a chat ID essentially represents a sub context.

We pass the context and chat ID info using the `context_name` and `chat_id` attributes on a `WiseAgentMessage`. There are some things we could consider:

* Do we want to rename `chat_id` to something like `sub_context_id` to make it more clear what it's used for?

* Do we want to make use of `WiseAgentContext` to represent a sub context? I haven't thought through what this would look like in detail yet but would require significant refactoring so that a `WiseAgentContext` could be associated with a list of `WiseAgentContexts` and then wouldn't make use of dictionaries. Basically for the 4 cases described above, a `WiseAgentContext` would be created where an ID is currently generated and so the information associated with that ID that is currently stored on the main context would be stored in the sub context. We'd need to think about if/how this would affect things like the message trace and participants.

I think we need to determine what we want the `WiseAgentContext` to represent.

Thoughts?

**Comment by maeste** [comment]
> I looked into our usage of context name and chat IDs.
Trying to answer inline to different points, even if the big discussion is probably at the end

> 
> Currently, chat IDs are generated and used for the following cases:
> 
> 1. Sequential coordination (an ID is generated for each request the coordinator receives and is used to keep track of the sequence of agents for the request and where to route the final response to)
> 2. Phased coordination (an ID is generated for each request the coordinator receives and is used to keep track of the agents required for each phase, where to route the final response to, and to keep track of the conversation history for the request)

This means that potentially sequential coordinator chat_id could be a sub-context of phased coordinator chat_id (in case a sequence of agents is the agent invoked in a phase)
That said,  the context is used to keep the collaboration type. So in this case have SeqCoord and PhasedCoord different context?

> 3. Assistant agent (a single ID is generated and used for all requests to keep track of the conversation history)
Almost the same consideration here. chat_id of any coordinator called from an Aisstant are sub-context of Assistant's chat_id

> 4. `LLMWiseAgentWithTools` (an ID is generated for each request the agent receives and is used to keep track of the tool calls needed for the request and the output of the tools).

Again the same
> 
> Currently, a `WiseAgentContext` has a name and can be associated with multiple requests so the context contains dictionaries that map chat IDs to various values.
But as said, it contains all the relevant information for coordination. So, any subcoordinator needs a subcontext. Am I wrong?
> 
> For example, if we have an `AssistantAgent` that can send a message to a `SequentialCoordinator` that invokes agents `A` and `B`, we'd have:
> 
> * a context name
> * a single chat ID generated by the `AssistantAgent`
> * a chat ID generated by the `SequentialCoordinator` to process each request it receives
> 
> If we have a `PassThroughClientAgent` that sends a message to a `SequentialCoordinator`, we'd have:
> 
> * a context name
> * a chat ID generated by the `SequentialCoordinator` to process each request it receives
> 

> Basically, we can see that a chat ID essentially represents a sub context.
> 

Yes and no. It contains at the moment different information. It makes sense to have it if in the same context we can have different chat history. I can see different cases for this. A possible one is when we would need to summarize the chat history for an agent using an LLM with limited tokens. Another one is when we would need to avoid to pass the full chat history to a specific agent (or a specific phase) to don't influence their results (for example in a challenging phase)

> We pass the context and chat ID info using the `context_name` and `chat_id` attributes on a `WiseAgentMessage`. There are some things we could consider:
> 
> * Do we want to rename `chat_id` to something like `sub_context_id` to make it more clear what it's used for?

As said it's not exactly a subcontext

> * Do we want to make use of `WiseAgentContext` to represent a sub context? I haven't thought through what this would look like in detail yet but would require significant refactoring so that a `WiseAgentContext` could be associated with a list of `WiseAgentContexts` and then wouldn't make use of dictionaries. Basically, for the 4 cases described above, a `WiseAgentContext` would be created where an ID is currently generated, and so the information associated with that ID that is currently stored on the main context would be stored in the subcontext. 

Hmm, Aren't we adding this sublevel in many cases artificially? It seems to me that we have more cases where ctx-chat_id relationship is 1.1, and just a few exceptions we have 1.n (and maybe sometimes wrongly). CAnnot be the info related to chat_id associated to the context? And just in case of real needs we can create subcontext?



> We'd need to think about if/how this would affect things like the message trace and participants.

Well, I'm opening another discussion about tracing. We have very poor information at the moment in the trace (basically just the WiseAgentMessage sent in each ctx). Definitely, having a better hierarchy of contexts would permit us to better represent what happened.

> 
> I think we need to determine what we want the `WiseAgentContext` to represent.
> 
> Thoughts?

I would be happy to discuss it more here, zulip, or call

Best
S.

**Reply by fjuma** [reply]
> > I looked into our usage of context name and chat IDs.
> > Trying to answer inline to different points, even if the big discussion is probably at the end
> 
> > Currently, chat IDs are generated and used for the following cases:
> > 
> > 1. Sequential coordination (an ID is generated for each request the coordinator receives and is used to keep track of the sequence of agents for the request and where to route the final response to)
> > 2. Phased coordination (an ID is generated for each request the coordinator receives and is used to keep track of the agents required for each phase, where to route the final response to, and to keep track of the conversation history for the request)
> 
> This means that potentially sequential coordinator chat_id could be a sub-context of phased coordinator chat_id (in case a sequence of agents is the agent invoked in a phase) That said, the context is used to keep the collaboration type. So in this case have SeqCoord and PhasedCoord different context?

The collaboration type is set for a chat_id on the context so currently, the chat_id for the sequential coordinator and the chat_id for the phased_coordinator would be associated with the same context and we'd know what type of collaboration each chat_id represents.


> Yes and no. It contains at the moment different information. It makes sense to have it if in the same context we can have different chat history. I can see different cases for this. A possible one is when we would need to summarize the chat history for an agent using an LLM with limited tokens. Another one is when we would need to avoid to pass the full chat history to a specific agent (or a specific phase) to don't influence their results (for example in a challenging phase)
> 
> > We pass the context and chat ID info using the `context_name` and `chat_id` attributes on a `WiseAgentMessage`. There are some things we could consider:
> > 
> > * Do we want to rename `chat_id` to something like `sub_context_id` to make it more clear what it's used for?
> 
> As said it's not exactly a subcontext
> 
> > * Do we want to make use of `WiseAgentContext` to represent a sub context? I haven't thought through what this would look like in detail yet but would require significant refactoring so that a `WiseAgentContext` could be associated with a list of `WiseAgentContexts` and then wouldn't make use of dictionaries. Basically, for the 4 cases described above, a `WiseAgentContext` would be created where an ID is currently generated, and so the information associated with that ID that is currently stored on the main context would be stored in the subcontext.
> 
> Hmm, Aren't we adding this sublevel in many cases artificially? It seems to me that we have more cases where ctx-chat_id relationship is 1.1, and just a few exceptions we have 1.n (and maybe sometimes wrongly). CAnnot be the info related to chat_id associated to the context? And just in case of real needs we can create subcontext?

I think using a `WiseAgentContext` to represent a sub context might be more complex than what we need.

The main case where a sub context will likely be needed is when we want to use a sequential coordinator from a phased coordinator. To keep track of things, we could ensure that the chat_id generated for the sequential coordinator takes the chat_id associated with the phased_coordinator and appends an _{new id} to it (as we discussed a bit last week). When WiseAgent.handle_response detects that sequential coordination is complete, it could then prune this ending part of the ID, detect that what we're left with is an id associated with phased coordination, and send the ACK response, update the conversation history, using this chat ID. We can discuss this more on our call today.

> 
> > We'd need to think about if/how this would affect things like the message trace and participants.
> 
> Well, I'm opening another discussion about tracing. We have very poor information at the moment in the trace (basically just the WiseAgentMessage sent in each ctx). Definitely, having a better hierarchy of contexts would permit us to better represent what happened.
> 
> > I think we need to determine what we want the `WiseAgentContext` to represent.
> > Thoughts?
> 
> I would be happy to discuss it more here, zulip, or call
> 
> Best S.
